File: src/memory/compressor.py
"""Context compression module - P0 priority.

Compresses conversation history to fit within token budget while
preserving key information needed for task continuity.
"""

from __future__ import annotations

import abc
from dataclasses import dataclass, field
from typing import Any


@dataclass
class Message:
    """A single conversation message."""

    role: str
    content: str
    token_count: int = 0
    importance: float = 1.0


@dataclass
class CompressionResult:
    """Result of context compression."""

    messages: list[Message]
    total_tokens: int
    original_tokens: int
    compression_ratio: float

    @property
    def tokens_saved(self) -> int:
        return self.original_tokens - self.total_tokens


class ContextCompressor(abc.ABC):
    """Abstract base class for context compression strategies.

    Context compression is the highest priority operation (P0).
    It must execute before any memory recall to ensure the current
    task stays on track within the token budget.
    """

    @abc.abstractmethod
    def compress(
        self, messages: list[Message], max_tokens: int
    ) -> CompressionResult:
        """Compress messages to fit within max_tokens.

        Args:
            messages: Full conversation history.
            max_tokens: Maximum allowed tokens after compression.

        Returns:
            CompressionResult with compressed messages.
        """
        ...

    @staticmethod
    def estimate_tokens(text: str) -> int:
        """Rough token estimation (4 chars ≈ 1 token)."""
        return len(text) // 4


class SlidingWindowCompressor(ContextCompressor):
    """Compression via sliding window with importance-based retention.

    Strategy:
    1. Always keep the system message and latest N messages.
    2. For older messages, retain only those above importance threshold.
    3. Summarize dropped messages into a single condensed message.
    """

    def __init__(
        self,
        recent_window: int = 5,
        importance_threshold: float = 0.5,
    ) -> None:
        self.recent_window = recent_window
        self.importance_threshold = importance_threshold

    def compress(
        self, messages: list[Message], max_tokens: int
    ) -> CompressionResult:
        if not messages:
            return CompressionResult(
                messages=[], total_tokens=0, original_tokens=0, compression_ratio=1.0
            )

        original_tokens = sum(m.token_count or self.estimate_tokens(m.content) for m in messages)

        # If already within budget, no compression needed
        if original_tokens <= max_tokens:
            return CompressionResult(
                messages=list(messages),
                total_tokens=original_tokens,
                original_tokens=original_tokens,
                compression_ratio=1.0,
            )

        # Split: system messages + recent window are protected
        system_msgs = [m for m in messages if m.role == "system"]
        non_system = [m for m in messages if m.role != "system"]

        recent = non_system[-self.recent_window :] if len(non_system) > self.recent_window else non_system
        older = non_system[: -self.recent_window] if len(non_system) > self.recent_window else []

        # Filter older messages by importance
        important_older = [m for m in older if m.importance >= self.importance_threshold]

        # Build summary of dropped messages
        dropped = [m for m in older if m.importance < self.importance_threshold]
        summary_parts: list[str] = []
        if dropped:
            summary_content = f"[Compressed {len(dropped)} earlier messages]"
            summary_msg = Message(
                role="system",
                content=summary_content,
                token_count=self.estimate_tokens(summary_content),
                importance=0.3,
            )
            summary_parts = [summary_msg]

        # Assemble final messages
        result_msgs = system_msgs + summary_parts + important_older + recent

        # Trim from important_older if still over budget
        total = sum(m.token_count or self.estimate_tokens(m.content) for m in result_msgs)
        while total > max_tokens and important_older:
            removed = important_older.pop(0)
            result_msgs = system_msgs + summary_parts + important_older + recent
            total = sum(m.token_count or self.estimate_tokens(m.content) for m in result_msgs)

        final_tokens = sum(m.token_count or self.estimate_tokens(m.content) for m in result_msgs)
        ratio = final_tokens / original_tokens if original_tokens > 0 else 1.0

        return CompressionResult(
            messages=result_msgs,
            total_tokens=final_tokens,
            original_tokens=original_tokens,
            compression_ratio=ratio,
        )

File: src/memory/compressor.ts
/**
 * Context compression module - P0 priority.
 *
 * Compresses conversation history to fit within token budget while
 * preserving key information needed for task continuity.
 */

export interface Message {
  role: string;
  content: string;
  tokenCount?: number;
  importance?: number;
}

export interface CompressionResult {
  messages: Message[];
  totalTokens: number;
  originalTokens: number;
  compressionRatio: number;
  tokensSaved: number;
}

/** Rough token estimation (4 chars ≈ 1 token). */
export function estimateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

/**
 * Abstract interface for context compression strategies.
 * Context compression is the highest priority operation (P0).
 */
export interface ContextCompressor {
  compress(messages: Message[], maxTokens: number): CompressionResult;
}

export interface SlidingWindowConfig {
  recentWindow?: number;
  importanceThreshold?: number;
}

/**
 * Compression via sliding window with importance-based retention.
 *
 * Strategy:
 * 1. Always keep the system message and latest N messages.
 * 2. For older messages, retain only those above importance threshold.
 * 3. Summarize dropped messages into a single condensed message.
 */
export class SlidingWindowCompressor implements ContextCompressor {
  private readonly recentWindow: number;
  private readonly importanceThreshold: number;

  constructor(config: SlidingWindowConfig = {}) {
    this.recentWindow = config.recentWindow ?? 5;
    this.importanceThreshold = config.importanceThreshold ?? 0.5;
  }

  compress(messages: Message[], maxTokens: number): CompressionResult {
    if (messages.length === 0) {
      return { messages: [], totalTokens: 0, originalTokens: 0, compressionRatio: 1.0, tokensSaved: 0 };
    }

    const getTokens = (m: Message) => m.tokenCount ?? estimateTokens(m.content);
    const originalTokens = messages.reduce((sum, m) => sum + getTokens(m), 0);

    // If already within budget, no compression needed
    if (originalTokens <= maxTokens) {
      return {
        messages: [...messages],
        totalTokens: originalTokens,
        originalTokens,
        compressionRatio: 1.0,
        tokensSaved: 0,
      };
    }

    // Split: system messages + recent window are protected
    const systemMsgs = messages.filter(m => m.role === 'system');
    const nonSystem = messages.filter(m => m.role !== 'system');

    const recent = nonSystem.length > this.recentWindow
      ? nonSystem.slice(-this.recentWindow)
      : [...nonSystem];
    const older = nonSystem.length > this.recentWindow
      ? nonSystem.slice(0, -this.recentWindow)
      : [];

    // Filter older messages by importance
    const importantOlder = older.filter(m => (m.importance ?? 1.0) >= this.importanceThreshold);
    const dropped = older.filter(m => (m.importance ?? 1.0) < this.importanceThreshold);

    // Build summary of dropped messages
    const summaryMsgs: Message[] = [];
    if (dropped.length > 0) {
      const summaryContent = `[Compressed ${dropped.length} earlier messages]`;
      summaryMsgs.push({
        role: 'system',
        content: summaryContent,
        tokenCount: estimateTokens(summaryContent),
        importance: 0.3,
      });
    }

    // Assemble final messages, trim from importantOlder if still over budget
    let resultMsgs = [...systemMsgs, ...summaryMsgs, ...importantOlder, ...recent];
    let total = resultMsgs.reduce((sum, m) => sum + getTokens(m), 0);

    const mutableImportant = [...importantOlder];
    while (total > maxTokens && mutableImportant.length > 0) {
      mutableImportant.shift();
      resultMsgs = [...systemMsgs, ...summaryMsgs, ...mutableImportant, ...recent];
      total = resultMsgs.reduce((sum, m) => sum + getTokens(m), 0);
    }

    const finalTokens = resultMsgs.reduce((sum, m) => sum + getTokens(m), 0);
    const ratio = originalTokens > 0 ? finalTokens / originalTokens : 1.0;

    return {
      messages: resultMsgs,
      totalTokens: finalTokens,
      originalTokens,
      compressionRatio: ratio,
      tokensSaved: originalTokens - finalTokens,
    };
  }
}
